src/transforms/pipelines/config.rs

```diff
@@ -42,65 +52,172 @@ impl Clone for PipelineConfig {
+        // future work.
+        if self.transforms.is_empty() {
+            return Err(format!("empty pipeline: {}", self.name).into());
+        }
+        let mut transforms = Vec::with_capacity(self.transforms.len());
+        for config in &self.transforms {
+            let transform = match config.build(ctx).await? {
+                Transform::Function(transform) => Box::new(transform),
+                Transform::Synchronous(transform) => transform,
+                _ => return Err(format!("non-sync transform in pipeline: {:?}", config).into()),
+            };
+            transforms.push(transform);
+        }
+
+        let buf_in = TransformOutputsBuf::new_with_capacity(
+            vec![Output::default(DataType::all())],
+            INTERIOR_BUFFER_SIZE,
+        );
+        let buf_out = buf_in.clone();
+        Ok(Transform::Synchronous(Box::new(Pipeline {
+            condition,
+            transforms,
+            buf_in,
+            buf_out,
+        })))
+    }
+
+    fn input(&self) -> Input {
+        if let Some(transform) = self.transforms.first() {
+            transform.input()
+        } else {
+            panic!("pipeline {} does not have transforms", self.name)
+        }
+    }
+
+    fn outputs(&self, schema: &schema::Definition) -> Vec<Output> {
+        if let Some(transform) = self.transforms.last() {
+            transform.outputs(schema)
+        } else {
+            panic!("pipeline {} does not have transforms", self.name)
+        }
+    }
+
+    fn transform_type(&self) -> &'static str {
+        // NOTE It'd be nice to avoid this boilerplate for transforms that are
+        // not meant to be directly user-configurable. In this case, we're
+        // really only implementing TransformConfig because we're ultimately
+        // returning this from an expansion, so it's likely this probably goes
+        // away when expansion goes away.
+        "pipeline"
+    }
+
+    fn enable_concurrency(&self) -> bool {
+        true
+    }
+}
+
 impl PipelineConfig {
     pub(super) fn expand(
         &mut self,
         name: &ComponentKey,
         inputs: &[String],
     ) -> crate::Result<Option<InnerTopology>> {
         let mut result = InnerTopology::default();
-        // define the name of the last output
-        let last_name = if self.transforms.is_empty() {
-            self.filter
-                .as_ref()
-                .map(|_filter| {
-                    let filter_name = name.join("filter");
-                    filter_name.join("success")
-                })
-                .ok_or_else(|| "mut have at least one transform or a filter".to_string())?
-        } else {
-            name.join(self.transforms.len() - 1)
-        };
+
+        result.inner.insert(
+            name.clone(),
+            InnerTopologyTransform {
+                inputs: inputs.to_vec(),
+                inner: Box::new(self.clone()),
+            },
+        );
+        // TODO: actually call outputs fn
```

Comment:
Is this still TODO?

Reply:
There's still some questions about outputs [here](https://github.com/vectordotdev/vector/pull/12602#discussion_r870821694) which I'll need some help to figure out -- how inputs and outputs get wired up eventually escapes me -- but I think this comment is dead now.